src/ai/backend/gateway/manager.py

```diff
@@ -0,0 +1,128 @@
+import asyncio
+import enum
+import functools
+import json
+import logging
+import sqlalchemy as sa
+
+from aiohttp import web
+
+from ai.backend.common.logging import BraceStyleAdapter
+
+from .exceptions import InvalidAPIParameters, ServerFrozen
+from ..manager.models import kernels, KernelStatus
+
+
+log = BraceStyleAdapter(logging.getLogger('ai.backend.gateway.manager'))
+
+
+class ManagerStatus(enum.Enum):
+    RUNNING = 'running'
+    FROZEN = 'frozen'
+
+    @classmethod
+    def is_member(cls, value):
+        return any(value == item.value for item in cls)
+
+
+def server_unfrozen_required(*args, **kwargs):
+
+    def wrapper(handler, gql=False):
+
+        @functools.wraps(handler)
+        async def _wrapped(request):
+            if gql:
+                # Check if server is frozen only for mutation queries.
+                try:
+                    params = await request.json()
+                    query = params['query']
+                    if not query.startswith('mutation'):
```

Comment:
This looks dangerous because a single GraphQL query may contain multiple queries and mutations in an arbitrary order.
Could we use graphene or its dependencies to check if the query actually contains any mutation?

Reply:
OK, I'll try to find related APIs!